app/Models/Loan.php

```diff
@@ -15,7 +16,43 @@ class Loan extends Model
     public const CURRENCY_SGD = 'SGD';
     public const CURRENCY_VND = 'VND';
 
-    use HasFactory;
+    use HasFactory, Notifiable;
+
+    public function repaid() {
+        $this->outstanding_amount = 0;
+        $this->status = Loan::STATUS_REPAID;
+    }
+
+    public function updateRepaid() {
+        if ($this->outstanding_amount == 0 || $this->status == Loan::STATUS_REPAID) {
+            $this->repaid();
+        }
+    }
+
+    protected static function booted()
+    {
+        static::creating(function (Loan $loan) {
+            $loan->outstanding_amount = $loan->outstanding_amount ?? $loan->amount;
+        });
+        static::updating(function (Loan $loan) {
+            $loan->updateRepaid();
+        });
+    }
+
+    public function setStatusAttribute($value) {
+        if ($value == Loan::STATUS_REPAID) {
+            $this->attributes['outstanding_amount'] = 0;
+        }
+        $this->attributes['status'] = $value;
+    }
+
+    public function setOutstandingAmountAttribute($value) {
+        assert($value >= 0);
```

Comment:
What would be a good way to do with such circumstances that are impossible normally? I suppose we want to terminate the control flow to prevent further consequences (e.g. setting `$value` to a negative number).

Reply:
I understand this assert here will block the whole flow after it when value is negative. 
But look into this circumstance, the value must not be negative and if it is, there must be a fault elsewhere maybe in the service where we do the calculation.
Then better to let the service handle this and make sure the value passing here always be positive.